// can-dom-events/dist/cjs/helpers/util.js
/*can-dom-events@1.3.2#helpers/util*/
'use strict';
var getCurrentDocument = require('can-globals/document/document');
var isBrowserWindow = require('can-globals/is-browser-window/is-browser-window');
function getTargetDocument(target) {
    return target.ownerDocument || getCurrentDocument();
}
function createEvent(target, eventData, bubbles, cancelable) {
    var doc = getTargetDocument(target);
    var event = doc.createEvent('HTMLEvents');
    var eventType;
    if (typeof eventData === 'string') {
        eventType = eventData;
    } else {
        eventType = eventData.type;
        for (var prop in eventData) {
            if (event[prop] === undefined) {
                event[prop] = eventData[prop];
            }
        }
    }
    if (bubbles === undefined) {
        bubbles = true;
    }
    event.initEvent(eventType, bubbles, cancelable);
    return event;
}
function isDomEventTarget(obj) {
    if (!(obj && obj.nodeName)) {
        return obj === window;
    }
    var nodeType = obj.nodeType;
    return nodeType === 1 || nodeType === 9 || nodeType === 11;
}
function addDomContext(context, args) {
    if (isDomEventTarget(context)) {
        args = Array.prototype.slice.call(args, 0);
        args.unshift(context);
    }
    return args;
}
function removeDomContext(context, args) {
    if (!isDomEventTarget(context)) {
        args = Array.prototype.slice.call(args, 0);
        context = args.shift();
    }
    return {
        context: context,
        args: args
    };
}
var fixSyntheticEventsOnDisabled = false;
(function () {
    if (!isBrowserWindow()) {
        return;
    }
    var testEventName = 'fix_synthetic_events_on_disabled_test';
    var input = document.createElement('input');
    input.disabled = true;
    var timer = setTimeout(function () {
        fixSyntheticEventsOnDisabled = true;
    }, 50);
    var onTest = function onTest() {
        clearTimeout(timer);
        input.removeEventListener(testEventName, onTest);
    };
    input.addEventListener(testEventName, onTest);
    try {
        var event = document.create('HTMLEvents');
        event.initEvent(testEventName, false);
        input.dispatchEvent(event);
    } catch (e) {
        onTest();
        fixSyntheticEventsOnDisabled = true;
    }
}());
function isDispatchingOnDisabled(element, event) {
    var eventType = event.type;
    var isInsertedOrRemoved = eventType === 'inserted' || eventType === 'removed';
    var isDisabled = !!element.disabled;
    return isInsertedOrRemoved && isDisabled;
}
function forceEnabledForDispatch(element, event) {
    return fixSyntheticEventsOnDisabled && isDispatchingOnDisabled(element, event);
}
module.exports = {
    createEvent: createEvent,
    addDomContext: addDomContext,
    removeDomContext: removeDomContext,
    isDomEventTarget: isDomEventTarget,
    getTargetDocument: getTargetDocument,
    forceEnabledForDispatch: forceEnabledForDispatch
};
